Extract sample dataset metadata in Cypress

nemanjaglumac opened the pull request:
### Status
READY

### What does this PR accomplish?
1. It dynamically extracts sample dataset metadata (`table` ids, and accompanied `table_field` ids) during snapshot creating and seeding phase
2. Stores them in a json file at the beginning of every Cypress run (thus, the file is ignored by Git)
3. Makes it available for use in every Cypress test (through `frontend/test/__support__/cypress_sample_dataset.js`) as `SAMPLE_DATASET` const

### Example of usage
```javascript
// before
import { foo, bar, withSampleDataset } from "__support__/cypress";

describe("some > description > block", () => {
  it("tests something", ()=> {
    withSampleDataset(({ ORDERS, ORDERS_ID })) => {
      someFunctionThatUses(ORDERS);
      const someObject = {
        baz: ORDERS_ID,
      },
    });
  });

  it("tests something else", ()=> {
    withSampleDataset(({ ORDERS, ORDERS_ID, PRODUCTS })) => {
      someFunctionThatUses(PRODUCTS);
      const someObject = {
        baz: ORDERS.CREATED_AT,
        buz: ORDERS_ID,
      },
    });
  });
});
```

```javascript
// now
import { foo, bar } from "__support__/cypress";
import { SAMPLE_DATASET } from "__support__/cypress_sample_dataset";

const { ORDERS, ORDERS_ID, PRODUCTS } = SAMPLE_DATASET;

describe("some > description > block", () => {
  it("tests something", ()=> {
    someFunctionThatUses(ORDERS);
    const someObject = {
      baz: ORDERS_ID,
    },
  });

  it("tests something else", ()=> {
    someFunctionThatUses(PRODUCTS);
    const someObject = {
      baz: ORDERS.CREATED_AT,
      buz: ORDERS_ID,
    },
  });
});
```

nemanjaglumac committed:
Extract `SAMPLE_DATASET` metadata

nemanjaglumac committed:
Test that metadata extraction is working properly in CI

nemanjaglumac committed:
Use absolute (aliased) path

nemanjaglumac commented:
CI errored with
```
/home/circleci/********/********/frontend/test/__support__/cypress.js
  4:43  error  Unable to resolve path to module './cypress-sample-dataset.json'  import/no-unresolved
```

I tried to use an absolute path (aliased one) - that one failed as well. 

My best guess is that it fails because linter is running first - before Cypress gets a chance to create sample dataset JSON from `frontend/test/snapshot-creators/default.cy.snap.js`

paulrosenzweig commented:
> My best guess is that it fails because linter is running first

If it's *only* the linter, you can add a comment to ignore that. However, the snapshot script will also import that file when it starts up. Maybe you could commit an empty file (or `{}`) that's overwritten with the real data? That will ensure the file can always be read and exported.

> Either that, or CI doesn't have permission to write files (which seems weird).

Snapshots rely on writing out SQL files for the backend, so that should be fine.

nemanjaglumac committed:
Ignore eslint `import/no-resolve` error for untracked test JSON file

nemanjaglumac commented:
> If it's _only_ the linter, you can add a comment to ignore that. However, the snapshot script will also import that file when it starts up. Maybe you could commit an empty file (or `{}`) that's overwritten with the real data? That will ensure the file can always be read and exported.

Before I saw your comment, I was searching for the ways to ignore this specific import and committed already :/. Let's see if that works first but maybe the idea with the empty file is cleaner. I just need to make sure it's committed first, but then I need to add it to the `.gitignore` afterwards.

nemanjaglumac committed:
Start with an empty `cypress-sample-dataset.json`

Note: This file will be overwritten every single time Cypress runs. It needs to be added to `.gitignore` in the next commit.

nemanjaglumac committed:
Add `cypress-sample-dataset.json` to gitignore

nemanjaglumac committed:
Update git index with `cy.exec()`

nemanjaglumac committed:
Use separate JS file for exporting sample dataset JSON

nemanjaglumac committed:
Remove cached JSON file

nemanjaglumac committed:
Update `.gitignore`

nemanjaglumac commented:
@paulrosenzweig I think this is ready for another inspection. I implemented your other suggestion and created another JS file that exports contents of this generated JSON file.

Everything seems to be working now. 🧙‍♂️

paulrosenzweig commented on frontend/test/metabase/scenarios/admin/settings/localization.cy.spec.js:
```diff
@@ -1,4 +1,7 @@
-import { restore, withSampleDataset, signInAsAdmin } from "__support__/cypress";
+import { restore, signInAsAdmin } from "__support__/cypress";
+import { SAMPLE_DATASET } from "__support__/cypress_sample_dataset";
+
+const { ORDERS, ORDERS_ID } = SAMPLE_DATASET;
```
🥳

paulrosenzweig replied:
Look at that! Access to the ids without nesting everything inside a callback! 🎉

nemanjaglumac replied:
@paulrosenzweig you mean [THIS](https://github.com/metabase/metabase/pull/14004) 🤓

paulrosenzweig reviewed (approved):
Looks great!
